### ISSUE TITLE
Review Back End

### ISSUE THREAD

**Issue by nopara73** [nopara73/MagicalCryptoWallet#2]
https://github.com/nopara73/MagicalCryptoWallet/blob/master/MagicalCryptoWallet.Backend/README.md

**Comment by lontivero** [comment]
I agree with most of the document that you shared here. 

Here are my observations:

# About the API requests

**POST broadcast** is okay. Just for the record, [here](https://en.bitcoin.it/wiki/Transaction_broadcasting) we can see how the rest of the world named this operation. Most of them call it pushtx, tx/send and broadcast. Sticking to a more rest-like scheme would require posting to /tx (POST /tx). 

**POST filters** I don't like it. I think we should use the GET verb if possible because it is simply, easy to use and invoke. I would suggest something like:
```
# Gets one specific filter
GET /api/filters/00000000000000000015d8bb1c4e15b1923bce24ad4c89d81ff930114c1fe416
```
Or,
```
# Gets multiple filters
GET /api/filters/00000000000000000015d8bb1c4e15b1923bce24ad4c89d81ff930114c1fe416,000000000000000000923bce24ad4c89d81ff930114c1f15d8bb1c4e15b1e416 
```
Also, 
```
# Gets multiple filters (from-to)
GET /api/filters?$from=00000000000000000015d8bb1c4e15b1923bce24ad4c89d81ff930114c1fe416&$to=000000000000000000923bce24ad4c89d81ff930114c1f15d8bb1c4e15b1e416 
```
This is particularly useful for initial (or periodical) synchronization where you can request from genesis to tip and the server can stream it to you in chunks or using continuations.

**POST fees**
This is IMO another candidate to use the GET http verb. Another observation is that I don't understand the ConfirmationTarget type, probably it is not obvious enough (what are C and E?)


# About the synchronization process

> When a client acquires a filter, it checks against its own keys and downloads the needed blocks from the Bitcoin P2P network, if needed.

The client needs to have more than just one source of filters, otherwise the server (the filters provider) can lie by creating fake, and specially crafted, filters that allow it to know the user's addresses. So, the client should download the filters from multiple sources and compare them. We could create a filter chain to help in the validation process and detect fake/broken filters.

**Comment by nopara73** [comment]
My problem with query string is that, its length is limited. And my problem with adding body to a GET is that some clients doesn't support it, eg. Postman, which I often use to test. This applies to both `fees` and `filters` requests. These are two practical problems, which may result in having an uglier API, but that uglier API would work better.

---

C: Conservative, E: Economical
`estimatesmartfee` takes these parameters, it's a relatively new command in Core. Maybe I should go with the full names, I wanted to save space, I'm not sure. 

---

> The client needs to have more than just one source of filters  
> So, the client should download the filters from multiple sources and compare them. 

That defeats the whole purpose of what we are doing with the back end.

> otherwise the server (the filters provider) can lie by creating fake, and specially crafted, filters that allow it to know the user's addresses

I don't see how that'd be the case. Please explain.  
The worst a server can do is to 1. make clients download blocks they are not interested in, 2. miss clients download blocks they are not interested in.  
But the client doesn't download blocks from the back end, so the back end won't know what the client is doing after that.  
If you are referring some kind of theoretical network attack, where the server hijacks all the full nodes the client is connected to, oh well, that's really the last thing to worry about, because 1. the server is already trusted with validation, 2. the assumption of most (if not all) anonymity networks, like Tor or I2P is that honest nodes outweigh malicious ones.

---

There is some confusion on the syncing process.  

> where you can request from genesis to tip 

You cannot request from the genesis. At initial sync, requesting starts from the first native segwit transaction's block.  

> filters?$from= to=

You don't know the hashes in front.  

> # Gets one specific filter, # Gets multiple filters

No, it doesn't. What you post is the filters that you already have and the following ones will be sent, it's to handle reorgs.  
Anyway, I just realized there are no need for multiple hashes, only one hash is fine, which means we can pass it in the query string as a parameter, so we can go with GET then. (This also simplifies the response slightly.)  

How a reorg would look like in this case:  

1. GET filter/hash -> filter not found  
2. Then the client tries it with its second best hash: GET filter/hash -> filter not found  
3. Third best hash GET filter/hash -> filters following the passed hash

**Comment by nopara73** [comment]
@lontivero I also made the `fees` query to be a GET, too, since it's unlikely that 1008 confirmation targets will be queried and I'm not even sure it'd fail.

**Comment by nopara73** [comment]
> I think there is no strong reason to save space in this particular case.

There is. When a blockchain analysis company looks at a transaction, fees are a reliable way to tell which wallet/service it was created with. I'm kind of puzzled on how to properly solve perfectly issue, but there are a couple of ways to mitigate it. https://github.com/nopara73/ZeroLink/#fee-rate-estimation
Furthermore we must protect our users against timing attacks of the server. It is possible that in the future we'll end up a bunch of fee requests. Not sure. Anyway, if we do and it'll be an issue we'll deal with it. For now, let's go with full names (`Economical` and `Conservative`.)

> I have to think a bit more about how to handle reorgs in an efficient way. 

If I can recall, there were 6 reorgs in the history of Bitcoin? No need to optimize for efficiency here.

> Firts of all, how to detect blockchain reorgs that could happen since the last time the wallet run.

Again, don't bother with it, I'm just a bad explainer, let me try again:  

1. Client: `GET filters/hash1`  
2. Server: Hey, I don't have this block, there must had been a reorg or you're just sending some bullshit hash.
3. Client:  `GET filters/hash1-1`  
4. Now, I have this hash, here is all the filters that I created after this block.

Or if you can understand it better from it, my swagger documentation and implementation is progressing, too:

![](https://i.imgur.com/FFXziUY.png)

> Also, if we receive more than one header for a parent block (there is a fork) we don't know which one will win and which one will be the orphans so, if we request the fiilters for those blocks to the server, it will serve them both.

There will always be one best block on the server, which it'll create filter with. If the block has been superseeded by another one, then the filter must be thrown out and continue with the real chain.
